pvlib/iam.py

```diff
@@ -527,3 +528,214 @@ def sapm(aoi, module, upper=None):
+              (89.5 <= zenith <= 90)
+            * 'ground': radiation reflected from the ground (zenith >= 90)
+
+        See [1]_ for a detailed description of each class.
+
+    num : int, optional
+        The number of increments in the zenith integration.
+        If not specified, N will follow the values used in [1]_:
+
+            * 'sky' or 'ground': num = 180
+            * 'horizon': num = 1800
+
+    Returns
+    -------
+    iam : numeric
+        AOI diffuse correction factor for the specified region.
+
+    See Also
+    --------
+    pvlib.iam.marion_diffuse
+
+    References
+    ----------
+    .. [1] B. Marion "Numerical method for angle-of-incidence correction
+       factors for diffuse radiation incident photovoltaic modules",
+       Solar Energy, Volume 147, Pages 344-348. 2017.
+       DOI: 10.1016/j.solener.2017.03.027
+
+    Examples
+    --------
+    >>> marion_integrate(pvlib.iam.ashrae, 20, 'sky')
+    0.9596085829811408
+
+    >>> from functools import partial
+    >>> f = partial(pvlib.iam.physical, n=1.3)
+    >>> marion_integrate(f, [20, 30], 'sky')
+    array([0.96225034, 0.9653219 ])
+    """
+
+    if num is None:
+        if region in ['sky', 'ground']:
+            num = 180
+        elif region == 'horizon':
+            num = 1800
+        else:
+            raise ValueError('Invalid region: {}'.format(region))
+
+    beta = np.radians(surface_tilt)
+    if isinstance(beta, pd.Series):
+        # convert Series to np array for broadcasting later
+        beta = beta.values
+    ai = np.pi/num  # angular increment
+
+    phi_range = np.linspace(0, np.pi, num, endpoint=False)
+    psi_range = np.linspace(0, 2*np.pi, 2*num, endpoint=False)
+
+    # the pseudocode in [1] do these checks at the end, but it's
+    # faster to do this criteria check up front instead of later.
+    if region == 'sky':
+        mask = phi_range + ai <= np.pi/2
+    elif region == 'horizon':
+        lo = 89.5 * np.pi/180
+        hi = np.pi/2
+        mask = (lo <= phi_range) & (phi_range + ai <= hi)
+    elif region == 'ground':
+        mask = (phi_range >= np.pi/2)
+    else:
+        raise ValueError('Invalid region: {}'.format(region))
+    phi_range = phi_range[mask]
+
+    # fast Cartesian product of phi and psi
+    angles = np.array(np.meshgrid(phi_range, psi_range)).T.reshape(-1, 2)
+    # index with single-element lists to maintain 2nd dimension so that
+    # these angle arrays broadcast across the beta array
+    phi_1 = angles[:, [0]]
+    psi_1 = angles[:, [1]]
+    phi_2 = phi_1 + ai
+    # psi_2 = psi_1 + ai  # not needed
+    phi_avg = phi_1 + 0.5*ai
+    psi_avg = psi_1 + 0.5*ai
+    term_1 = np.cos(beta) * np.cos(phi_avg)
+    # we are free to choose any gamma between zero and 2pi, so choose zero
+    # and omit it from the cos(psi_avg) term
+    term_2 = np.sin(beta) * np.sin(phi_avg) * np.cos(psi_avg)
+    aoi = np.arccos(term_1 + term_2)
```

Comment:
Can you save `term1 + term2` so you don't have to back-calculated it on line 720 below?
_EG_:
```python
cosaoi = term1 + term2
aoi = arcos(cosaoi)
```
then below on L720:
```python
cosaoi_dAs = cosaoi * dAs
```

Also, Bill's paper has a gamma term for the azimuth of the panels, I feel like you mentioned that up above, but I'm forgetting, why isn't it included here?
```python
cosaoi = term1 + term2 - gamma_pvmodule_azimuth
```
sorry, I didn't read through the previous comments, so forgive me if redundant

Reply:
+1 for saving `cosaoi`, I'll make that change.

The panel azimuth term is omitted because the system is rotationally invariant -- it's all about isotropic diffuse irradiance, so the panel azimuth can be chosen freely, and I chose to make it zero and leave it out as a tiny speed optimization.  Bill sets gamma to pi in the Section 3 pseudocode but mentions that gamma can be set to any value between 0 and 2pi.  I'll expand that explanatory comment to hopefully make that more clear.